### DISCUSSION TITLE
Compatibility - For Dummies?

### DISCUSSION THREAD

**Discussion by lloy0076** [mozilla/rhino#1257] [General]
The table(s) at https://mozilla.github.io/rhino/compat/engines.html have far more **RED** or even **ERROR** than green, and even then a lot of **YELLOW**.

Is there a "compatiblity for dummies" buried somewhere? Maybe something like, "If you set your babel preset to XXX"?

I'm not entirely certain precisely what I'm asking; I know that could be refined.

I know there are "per centage compatibles" on that table, but - to take an example - if the `=` (assignment) operator really did not work, then even if it were 90% compatible it would be somewhat difficult to use.

**Comment by p-bakker** [comment]
Hi, this is kind of a difficult question to answer and it depends what you're looking for.

If you're looking for compatibility with the latest EcmaScript standard, then Rhino leaves quite a few things to be desired.
If you're looking for general JavaScript functionality to script some logic, then Rhino is quite compatible and you're not likely to run into any JavaScript incompatibilities.

As for the most notable incompatibilities and missing features:
- Rhino doesn't properly support strict mode
- Rhino's (block) scoping of let, const and functions isn't 100% compatible
- Rhino doesn't support async functions
- Rino doesn't support EcmaScript Modules (EMS), but does support CommonJS

You can look at all the open cases in the EXxxxx milestones in the repo for more a exhaustive & detailed list

As of around the Rhino 1.7.14 release, there's also a rhino target in Babel: https://babeljs.io/docs/en/options#targets

Hope this helps!

**Comment by lloy0076** [comment]
Some broad statements such as the above might be reassuring - with the amount of "doesn't work" it'd be easy to conclude - falsely - that Rhino is just not useful in current times.

**Comment by p-bakker** [comment]
Not exactly sure how to interpret your comment, but, as with any technology, anyone looking at potentially using Rhino needs to consider whether it fits their needs. 

Rhino is used on a daily basis by many applications, serving many, many users, so it's usefulness is definitely there

But if you're looking for all the latest EcmaScript features, Rhino atm isn't the best fit. The aim is to improve EcmaScript compliance in Rhino, but progress depends on volunteers and the community.
